src/main/java/org/zalando/nakadi/service/timeline/TimelineSyncImpl.java

```diff
@@ -0,0 +1,333 @@
+        final Optional<Long> expectedFinish = Optional.ofNullable(timeoutMs).map(t -> System.currentTimeMillis() + t);
+        final AtomicInteger versionToWait = new AtomicInteger();
+        runLocked(() -> {
+            try {
+                final int latestVersion = readData("/version", null, Integer::valueOf);
+                versionToWait.set(latestVersion + 1);
+                zooKeeperHolder.get().setData().forPath(
+                        toZkPath("/version"), String.valueOf(versionToWait.get()).getBytes(Charsets.UTF_8));
+                LOG.info("Wrote to ZK version to wait (was: {}), new one: {}", latestVersion, versionToWait.get());
+            } catch (final RuntimeException ex) {
+                throw ex;
+            } catch (final Exception e) {
+                throw new RuntimeException(e);
+            }
+        });
+        // Wait for all nodes to have latest version.
+        final AtomicBoolean latestVersionIsThere = new AtomicBoolean(false);
+        while (!latestVersionIsThere.get()) {
+            LOG.info("Waiting for all nodes to have the same version {}", versionToWait.get());
+            if (expectedFinish.isPresent() && System.currentTimeMillis() > expectedFinish.get()) {
+                throw new RuntimeException("Timed out while updating version to " + versionToWait.get());
+            }
+            Thread.sleep(TimeUnit.SECONDS.toMillis(1));
+            runLocked(() -> {
+                try {
+                    final List<String> nodes = zooKeeperHolder.get().getChildren().forPath(toZkPath("/nodes"));
+                    boolean allOk = true;
+                    for (final String node : nodes) {
+                        final int nodeVersion = readData("/nodes/" + node, null, Integer::valueOf);
+                        final boolean nodeOk = nodeVersion >= versionToWait.get();
+                        if (!nodeOk) {
+                            allOk = false;
+                        }
+                        LOG.info("Node {} OK (current: {}, expected: {})", node, nodeVersion, versionToWait.get());
+                    }
+                    if (allOk) {
+                        latestVersionIsThere.set(true);
+                    }
+                } catch (final RuntimeException e) {
+                    throw e;
+                } catch (final Exception ex) {
+                    throw new RuntimeException(ex);
+                }
+            });
+        }
+        LOG.info("Version update to {} complete", versionToWait.get());
+    }
+
+    @Override
+    public void startTimelineUpdate(final String eventType, final long timeoutMs) throws InterruptedException {
+        LOG.info("Starting timeline update for event type {} with timeout {} ms", eventType, timeoutMs);
+        final String etZkPath = toZkPath("/locked_et/" + eventType);
+        boolean successful = false;
+        try {
+            zooKeeperHolder.get().create().withMode(CreateMode.EPHEMERAL)
```

Comment:
@rcillo Anyway, I added additional exception to treat this information separately